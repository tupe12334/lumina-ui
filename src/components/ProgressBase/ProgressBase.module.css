.progressBar {
  position: relative;
  width: 100%;
  background-color: #f3f4f6;
  border-radius: 0.375rem;
  overflow: hidden;
}

.progressFill {
  height: 100%;
  background-color: #3b82f6;
  border-radius: inherit;
  transition: width 0.3s ease-in-out;
}

.progressText {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 0.75rem;
  font-weight: 500;
  color: #374151;
}

/* Size variants */
.small {
  height: 0.5rem;
}

.medium {
  height: 0.75rem;
}

.large {
  height: 1rem;
}

/* Color variants */
.default .progressFill {
  background-color: #3b82f6;
}

.success .progressFill {
  background-color: #10b981;
}

.error .progressFill {
  background-color: #ef4444;
}

.warning .progressFill {
  background-color: #f59e0b;
}