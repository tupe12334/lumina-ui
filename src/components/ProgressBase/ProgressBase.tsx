import React from 'react'
import { clsx } from 'clsx'
import styles from './ProgressBase.module.css'

export interface ProgressBaseProps {
  value: number
  min?: number
  max?: number
  variant?: 'default' | 'success' | 'error' | 'warning'
  size?: 'small' | 'medium' | 'large'
  className?: string
  showValue?: boolean
  'aria-label'?: string
  'aria-labelledby'?: string
}

export const ProgressBase: React.FC<ProgressBaseProps> = (props) => {
  const {
    value,
    min,
    max,
    variant,
    size,
    className,
    showValue,
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledBy,
  } = props

  const minValue = min !== undefined ? min : 0
  const maxValue = max !== undefined ? max : 100
  const variantValue = variant !== undefined ? variant : 'default'
  const sizeValue = size !== undefined ? size : 'medium'
  const showValueFlag = showValue !== undefined ? showValue : false

  const clampedValue = Math.min(Math.max(value, minValue), maxValue)
  const percentage = ((clampedValue - minValue) / (maxValue - minValue)) * 100

  const getVariantClass = () => {
    if (variantValue === 'success') return styles.success
    if (variantValue === 'error') return styles.error
    if (variantValue === 'warning') return styles.warning
    return styles.default
  }

  const getSizeClass = () => {
    if (sizeValue === 'small') return styles.small
    if (sizeValue === 'large') return styles.large
    return styles.medium
  }

  return (
    <div
      className={clsx(
        styles.progressBar,
        getVariantClass(),
        getSizeClass(),
        className
      )}
      role="progressbar"
      aria-valuenow={clampedValue}
      aria-valuemin={minValue}
      aria-valuemax={maxValue}
      aria-label={ariaLabel}
      aria-labelledby={ariaLabelledBy}
    >
      <div
        className={styles.progressFill}
        style={{ width: `${percentage}%` }}
      />
      {showValueFlag && (
        <span className={styles.progressText} aria-hidden="true">
          {Math.round(percentage)}%
        </span>
      )}
    </div>
  )
}