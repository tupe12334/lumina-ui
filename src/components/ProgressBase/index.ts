export { ProgressBase, type ProgressBaseProps } from './ProgressBase'