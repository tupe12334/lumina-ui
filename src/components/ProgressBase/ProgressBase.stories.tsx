import type { Meta, StoryObj } from '@storybook/react'
import { ProgressBase } from './ProgressBase'

const meta: Meta<typeof ProgressBase> = {
  title: 'Components/ProgressBase',
  component: ProgressBase,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    value: {
      control: { type: 'range', min: 0, max: 100, step: 1 },
    },
    variant: {
      control: 'select',
      options: ['default', 'success', 'error', 'warning'],
    },
    size: {
      control: 'select',
      options: ['small', 'medium', 'large'],
    },
  },
}

export default meta
type Story = StoryObj<typeof ProgressBase>

export const Default: Story = {
  args: {
    value: 50,
  },
}

export const Success: Story = {
  args: {
    value: 75,
    variant: 'success',
  },
}

export const Error: Story = {
  args: {
    value: 30,
    variant: 'error',
  },
}

export const Warning: Story = {
  args: {
    value: 60,
    variant: 'warning',
  },
}

export const Small: Story = {
  args: {
    value: 40,
    size: 'small',
  },
}

export const Medium: Story = {
  args: {
    value: 40,
    size: 'medium',
  },
}

export const Large: Story = {
  args: {
    value: 40,
    size: 'large',
  },
}

export const WithValue: Story = {
  args: {
    value: 65,
    showValue: true,
  },
}

export const Complete: Story = {
  args: {
    value: 100,
    variant: 'success',
  },
}

export const Empty: Story = {
  args: {
    value: 0,
  },
}