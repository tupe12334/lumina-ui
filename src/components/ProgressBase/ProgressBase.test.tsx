import { render, screen } from '@testing-library/react'
import { ProgressBase } from './ProgressBase'

describe('ProgressBase', () => {
  test('renders correctly with default props', () => {
    const { container } = render(<ProgressBase value={50} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('displays correct progress value', () => {
    render(<ProgressBase value={75} />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-valuenow', '75')
    expect(progressBar).toHaveAttribute('aria-valuemin', '0')
    expect(progressBar).toHaveAttribute('aria-valuemax', '100')
  })

  test('clamps value to min and max bounds', () => {
    const { rerender } = render(<ProgressBase value={150} min={0} max={100} />)

    let progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-valuenow', '100')

    rerender(<ProgressBase value={-10} min={0} max={100} />)
    progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-valuenow', '0')
  })

  test('applies correct CSS classes for variants', () => {
    const { container } = render(<ProgressBase value={50} variant="success" />)
    const progressBar = container.firstChild
    if (progressBar && 'className' in progressBar) {
      expect(progressBar.className).toMatch(/success/)
    }
  })

  test('applies correct CSS classes for sizes', () => {
    const { container } = render(<ProgressBase value={50} size="large" />)
    const progressBar = container.firstChild
    if (progressBar && 'className' in progressBar) {
      expect(progressBar.className).toMatch(/large/)
    }
  })

  test('applies custom className', () => {
    const { container } = render(<ProgressBase value={50} className="custom-class" />)
    const progressBar = container.firstChild
    expect(progressBar).toHaveClass('custom-class')
  })

  test('shows percentage text when showValue is true', () => {
    render(<ProgressBase value={75} showValue={true} />)

    expect(screen.getByText('75%')).toBeInTheDocument()
  })

  test('hides percentage text when showValue is false', () => {
    render(<ProgressBase value={75} showValue={false} />)

    expect(screen.queryByText('75%')).not.toBeInTheDocument()
  })

  test('supports custom min and max values', () => {
    render(<ProgressBase value={15} min={10} max={20} />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-valuenow', '15')
    expect(progressBar).toHaveAttribute('aria-valuemin', '10')
    expect(progressBar).toHaveAttribute('aria-valuemax', '20')
  })

  test('calculates percentage correctly with custom range', () => {
    render(<ProgressBase value={15} min={10} max={20} showValue={true} />)

    // 15 in range 10-20 should be 50%
    expect(screen.getByText('50%')).toBeInTheDocument()
  })

  test('supports aria-label', () => {
    render(<ProgressBase value={50} aria-label="Upload progress" />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-label', 'Upload progress')
  })

  test('supports aria-labelledby', () => {
    render(<ProgressBase value={50} aria-labelledby="progress-label" />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-labelledby', 'progress-label')
  })

  test('progress fill has correct width style', () => {
    const { container } = render(<ProgressBase value={60} />)
    const progressFill = container.querySelector('[style*="width: 60%"]')
    expect(progressFill).toBeInTheDocument()
  })
})