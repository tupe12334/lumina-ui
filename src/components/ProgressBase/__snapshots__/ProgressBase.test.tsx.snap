// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressBase > renders correctly with default props 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  class="_progressBar_1931c0 _default_1931c0 _medium_1931c0"
  role="progressbar"
>
  <div
    class="_progressFill_1931c0"
    style="width: 50%;"
  />
</div>
`;
